### PR TITLE
[New Resource] azure_data_factory_v2 (again)

### PR DESCRIPTION
**Pull request by hbuckle**
Resubmitting [this](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1773) PR as I no longer have access to the original branch.

Github repo association is working correctly now - you are prompted to allow access to the repo when you log in to the data factory UI for the first time.

### CODE REVIEW

File: azurerm/resource_arm_data_factory_v2.go [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+	if repoType == datafactory.TypeFactoryRepoConfiguration {
+		d.Set("vsts_configuration", repo)
+		d.Set("github_configuration", repo)
+	}
+
+	if err := d.Set("identity", flattenArmDataFactoryV2Identity(resp.Identity)); err != nil {
+		return fmt.Errorf("Error setting Data Factory %q `identity` (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	if tags := resp.Tags; tags != nil {
```
Comment (katbyte) [inline azurerm/resource_arm_data_factory_v2.go]: We don't need this check as `flattenAndSetTags` will correctly handle a `nil`

File: azurerm/resource_arm_data_factory_v2.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+		return fmt.Errorf("Error making Read request on AzureRM Data Factory %s: %+v", name, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resourceGroup)
+	if location := resp.Location; location != nil {
+		d.Set("location", azureRMNormalizeLocation(*location))
+	}
+
+	repoType, repo := flattenArmDataFactoryV2RepoConfiguration(&resp)
```
Comment (katbyte) [inline azurerm/resource_arm_data_factory_v2.go]: Could we break this out into two separate functions? one for github and one for vsts

File: azurerm/resource_arm_data_factory_v2_test.go [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+	config := testAccAzureRMDataFactoryV2_tags(ri, testLocation())
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckAzureRMDataFactoryV2Destroy,
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+			},
```
Comment (katbyte) [inline azurerm/resource_arm_data_factory_v2_test.go]: Could we add some checks here?

File: azurerm/config.go [hunk 4]
```diff
@@ -44,7 +45,7 @@ import (
 	"github.com/Azure/azure-sdk-for-go/services/preview/msi/mgmt/2015-08-31-preview/msi"
 	"github.com/Azure/azure-sdk-for-go/services/preview/operationalinsights/mgmt/2015-11-01-preview/operationalinsights"
 	"github.com/Azure/azure-sdk-for-go/services/preview/operationsmanagement/mgmt/2015-11-01-preview/operationsmanagement"
-	"github.com/Azure/azure-sdk-for-go/services/preview/resources/mgmt/2018-03-01-preview/management"
+	managementgroups "github.com/Azure/azure-sdk-for-go/services/preview/resources/mgmt/2018-03-01-preview/management"
```
[the remark this answers is not in the document]
Comment (hbuckle) [reply azurerm/config.go]: It must have been added by one of the VS Code linters. I've removed it anyway

File: azurerm/config.go [hunk 5]
```diff
@@ -68,7 +69,7 @@ import (
 	"github.com/Azure/go-autorest/autorest/adal"
 	az "github.com/Azure/go-autorest/autorest/azure"
 	"github.com/hashicorp/go-azure-helpers/authentication"
-	"github.com/hashicorp/go-uuid"
+	uuid "github.com/hashicorp/go-uuid"
```
[the remark this answers is not in the document]
Comment (hbuckle) [reply azurerm/config.go]: It must have been added by one of the VS Code linters. I've removed it anyway

File: azurerm/resource_arm_data_factory_v2.go [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+						},
+						"tenant_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"github_configuration": {
```
[the remark this answers is not in the document]
Comment (hbuckle) [reply azurerm/resource_arm_data_factory_v2.go]: It already does, line 100

File: azurerm/resource_arm_data_factory_v2.go [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+						"root_folder": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+				ConflictsWith: []string{"vsts_configuration"},
+			},
+
+			"vsts_configuration": {
```
[the remark this answers is not in the document]
Comment (hbuckle) [reply azurerm/resource_arm_data_factory_v2.go]: It already does, line 141

File: azurerm/resource_arm_data_factory_v2.go (hunk 2 above)
Comment (hbuckle) [reply azurerm/resource_arm_data_factory_v2.go]: Any particular reason? It's kind of tricky as you have to drill down into the properties before you even find out what repo type you are dealing with.
Happy to make the change if there's a better way to do it

File: azurerm/provider.go [hunk 8]
```diff
@@ -156,6 +156,7 @@ func Provider() terraform.ResourceProvider {
 			"azurerm_container_group":                        resourceArmContainerGroup(),
 			"azurerm_cosmosdb_account":                       resourceArmCosmosDBAccount(),
 			"azurerm_databricks_workspace":                   resourceArmDatabricksWorkspace(),
+			"azurerm_data_factory_v2":                        resourceArmDataFactoryV2(),
```
[the remark this answers is not in the document]
Comment (katbyte) [reply azurerm/provider.go]: Looking at the portal there doesn't seem to be much difference between v1 or v2. Are the APIs different/how dos one differ between the two?

[the remark this answers is not in the document]
Comment (hbuckle) [reply azurerm/provider.go]: They do use slightly different APIs [v1](https://docs.microsoft.com/en-us/rest/api/datafactory/v1/data-factory-data-factory), [v2](https://docs.microsoft.com/en-us/rest/api/datafactory/factories/createorupdate)

[the remark this answers is not in the document]
Comment (hbuckle) [reply azurerm/provider.go]: Azure Powershell also differentiates between v1 and v2 https://docs.microsoft.com/en-us/powershell/module/az.datafactory/?view=azps-1.1.0#data_factories

**Comment by mbfrahry** [comment]
Hey @hbuckle, apologies for the time its taken on this PR. We've been having some internal discussions about the naming for this resource and we believe that removing `v2` is probably for the best because it seems like Microsoft is pushing v2 as the standard. Once that's taken care of, we can move this into the provider. Thank you again for taking the time to write this resource out.

**Comment by mbfrahry** [comment]
Hey @hbuckle, just checking in one more time. Is it ok if I start updating your PR so we can get it in to the provider?

**Comment by tombuildsstuff** [comment]
hey @hbuckle 

Apologies for the delayed review here!

Since we've not heard back from you here so that we can get this merged @mbfrahry has added some extra commits to resolve the comments above; I hope you don't mind - which has resulted in #3159 - rather than having multiple PR's open tracking the same feature - I'm going to close this PR in favour of #3159.

Thanks!

**Comment by hbuckle** [comment]
@mbfrahry @mbfrahry - no problem, I'd kind of forgotten about this one, thanks for finishing it off

**Pull request closed without merging**